cpp/src/arrow/types/struct.cc

```diff
@@ -17,4 +17,68 @@
 
 #include "arrow/types/struct.h"
 
-namespace arrow {}  // namespace arrow
+#include <sstream>
+
+namespace arrow {
+
+bool StructArray::EqualsExact(const StructArray& other) const {
+  if (this == &other) { return true; }
+  if (field_arrays_.size() != other.fields().size()) { return false; }
+  if (null_count_ != other.null_count_) { return false; }
+
+  if (null_count_ > 0) {
+    bool equal_null_bitmap =
+        null_bitmap_->Equals(*other.null_bitmap_, util::bytes_for_bits(length_));
+    if (!equal_null_bitmap) { return false; }
+
+    bool fields_equal = true;
+    for (size_t i = 0; i < field_arrays_.size(); ++i) {
+      fields_equal = field_arrays_[i].get()->Equals(other.field(i));
```

Comment:
@wesm This is the intention of the RangeEquals method.   The problem with accepting the entire mask bitmap, is nested struct types would require a memory allocation to calculate the logical conjunction of the bitmask to pass to through to the top level struct's grandchildren arrays.   This could be done but would would make for an awkward contract of having to pass memory pool in some form through equality method (or keep one around on each array).

The original API I was thinking of was to have a RangeEquals take int64_t (or possibly larger SIMD type) bitmap that could be used to do the the conjunction.  But I thought the contract for that method was also a little bit awkward as well/hard to explain (and require a more complicated implementation up front).